tests/run/posonly.pyx

```diff
@@ -0,0 +1,323 @@
+    >>> test_positional_only_and_arg_invalid_calls(1)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_and_arg_invalid_calls() takes exactly 3 positional arguments (1 given)
+    >>> test_positional_only_and_arg_invalid_calls(1,2,3,4)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_and_arg_invalid_calls() takes exactly 3 positional arguments (4 given)
+    """
+    pass
+
+def test_positional_only_and_optional_arg_invalid_calls(a, b, /, c=3):
+    """
+    >>> test_positional_only_and_optional_arg_invalid_calls(1, 2)
+    >>> test_positional_only_and_optional_arg_invalid_calls(1)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_and_optional_arg_invalid_calls() takes at least 2 positional arguments (1 given)
+    >>> test_positional_only_and_optional_arg_invalid_calls()
+    Traceback (most recent call last):
+    TypeError: test_positional_only_and_optional_arg_invalid_calls() takes at least 2 positional arguments (0 given)
+    >>> test_positional_only_and_optional_arg_invalid_calls(1, 2, 3, 4)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_and_optional_arg_invalid_calls() takes at most 3 positional arguments (4 given)
+    """
+    pass
+
+def test_positional_only_invalid_calls(a, b, /):
+    """
+    >>> test_positional_only_invalid_calls(1, 2)
+    >>> test_positional_only_invalid_calls(1)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_invalid_calls() takes exactly 2 positional arguments (1 given)
+    >>> test_positional_only_invalid_calls()
+    Traceback (most recent call last):
+    TypeError: test_positional_only_invalid_calls() takes exactly 2 positional arguments (0 given)
+    >>> test_positional_only_invalid_calls(1, 2, 3)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_invalid_calls() takes exactly 2 positional arguments (3 given)
+    """
+    pass
+
+def test_positional_only_with_optional_invalid_calls(a, b=2, /):
+    """
+    >>> test_positional_only_with_optional_invalid_calls(1)
+    >>> test_positional_only_with_optional_invalid_calls()
+    Traceback (most recent call last):
+    TypeError: test_positional_only_with_optional_invalid_calls() takes at least 1 positional argument (0 given)
+    >>> test_positional_only_with_optional_invalid_calls(1, 2, 3)
+    Traceback (most recent call last):
+    TypeError: test_positional_only_with_optional_invalid_calls() takes at most 2 positional arguments (3 given)
+    """
+    pass
+
+def test_no_standard_args_usage(a, b, /, *, c):
+    """
+    >>> test_no_standard_args_usage(1, 2, c=3)
+    >>> test_no_standard_args_usage(1, b=2, c=3)
+    Traceback (most recent call last):
+    TypeError: test_no_standard_args_usage() takes exactly 2 positional arguments (1 given)
+    """
+    pass
+
+#def test_change_default_pos_only():
+# TODO: probably remove this, since we have no __defaults__ in Cython?
```

Comment:
Good to know, done.  Some of these are tests are still commented out for two reasons:
(1) In the CPython tests, __defaults__ is modified during the test and this affects later calls to the function.  Is there a way to enable this feature (writability of __defaults__) in Cython?
(2) The tests refer to the member `f.__code__.co_posonlyargcount`.  This is a new addition to PyCodeObject as part of the posonly change.  Since the posonly change has yet to be merged into CPython, this member is missing currently.

Reply:
1) I think writing to `__defaults__` is a separate feature. Not worth testing here.
2) Sounds like a `hasattr()` might help. We are testing a CPython feature here, after all, not really a Cython feature. It's more like "if CPython supports it, then Cython should do the right thing. If not, then that's fine, too".